igvc_navigation/src/mapper/map_utils.cpp

```diff
@@ -0,0 +1,231 @@
+
+  PointCloud::Ptr cloud_filtered = raw_pc.makeShared();
+
+  // Create filtering object
+  pcl::ExtractIndices<pcl::PointXYZ> extract;
+
+  seg.setInputCloud(cloud_filtered);
+  seg.segment(*inliers, *coefficients);
+
+  if (inliers->indices.empty())
+  {
+    return std::nullopt;
+  }
+  else
+  {
+    ROS_DEBUG("Ground plane found: %zu/%zu inliers. Coeff: %f %f %f %f", inliers->indices.size(),
+              cloud_filtered->size(), coefficients->values.at(0), coefficients->values.at(1),
+              coefficients->values.at(2), coefficients->values.at(3));
+    extract.setInputCloud(cloud_filtered);
+    extract.setIndices(inliers);
+    extract.setNegative(false);
+    extract.filter(ground);
+
+    // remove ground points from full pointcloud
+    if (inliers->indices.size() != cloud_filtered->size())
+    {
+      extract.setNegative(true);
+      PointCloud out;
+      extract.filter(out);
+      nonground += out;
+      *cloud_filtered = out;
+    }
+
+    return GroundPlane{ coefficients->values.at(0), coefficients->values.at(1), coefficients->values.at(2),
+                        coefficients->values.at(3) };
+  }
+}
+
+void fallbackFilter(const PointCloud& raw_pc, PointCloud& ground, PointCloud& nonground, FallbackOptions options)
+{
+  pcl::PassThrough<pcl::PointXYZ> fallback;
+  fallback.setFilterFieldName("z");
+  fallback.setFilterLimits(-options.plane_distance, options.plane_distance);
+  fallback.filter(ground);
+
+  fallback.setFilterLimitsNegative(true);
+  fallback.filter(nonground);
+}
+
+void blur(cv::Mat& blurred_map, double kernel_size)
+{
+  cv::Mat original = blurred_map.clone();
+  cv::blur(blurred_map, blurred_map, cv::Size(kernel_size, kernel_size));
+  cv::max(original, blurred_map, blurred_map);
+}
+
+void getEmptyPoints(const pcl::PointCloud<pcl::PointXYZ>& pc, pcl::PointCloud<pcl::PointXYZ>& empty_pc,
+                    double angular_resolution, EmptyFilterOptions options)
+{
+  // Iterate over pointcloud, insert discretized angles into set
+  std::unordered_set<int> discretized_angles{};
+  for (auto i : pc)
+  {
+    double angle = atan2(i.y, i.x);
+    discretized_angles.emplace(MapUtils::discretize(angle, angular_resolution));
+  }
+
+  // For each angle, if it's not in the set (empty), put it into a pointcloud
+  static double coeff = 1 / angular_resolution;
+  // From Robot's frame. Need to rotate angle to world frame
+  for (int i = static_cast<int>(options.start_angle * coeff); i < options.end_angle * coeff; i++)
+  {
+    if (discretized_angles.find(i) == discretized_angles.end())
+    {
+      double angle = i * angular_resolution;
+      pcl::PointXYZ point{ static_cast<float>(options.miss_cast_distance * cos(angle)),
+                           static_cast<float>(options.miss_cast_distance * sin(angle)), 0 };
+      empty_pc.points.emplace_back(point);
+    }
+  }
+}
+
+void projectToPlane(PointCloud& projected_pc, const GroundPlane& ground_plane, const cv::Mat& image,
```

Comment:
Make sure this method projects to the "correct" ground plane. That is, z=0 as defined by base_link. I haven't viewed the rest of the code yet but i'm adding this in just in case.